File: nano-banana-compare/google-flow-nano-banana-compare.py
"""

Script version 1.1, September 11, 2026

Batch-compare Google Flow's Nano Banana image models (2 Lite, 2, Pro) with the useapi.net API.
Reads prompts.json (one entry per model), submits each to the synchronous POST /images endpoint,
and downloads every returned image with a model-labeled filename.
For more details visit https://useapi.net/docs/api-google-flow-v1/post-google-flow-images

Companion tutorial:
  https://useapi.net/docs/articles/google-flow-nano-banana-compare

Installation Instructions:
==========================

You need Python 3.x installed to run this script (standard library only, no dependencies):

- Windows, macOS, Linux: https://www.python.org/

Verify with:

   python3 --version

Running the Script:
===================

Usage: python3 google-flow-nano-banana-compare.py <API_TOKEN> <EMAIL> [PROMPTS_FILE]

Replace API_TOKEN with your actual useapi.net API token, see https://useapi.net/docs/start-here/setup-useapi
Replace EMAIL with configured Google Flow email account, see https://useapi.net/docs/start-here/setup-google-flow
If optional PROMPTS_FILE not provided prompts.json will be used.

Example:
--------

python3 google-flow-nano-banana-compare.py user:1234-abcdefhijklmnopqrstuv my@email.com

Changelog:
==========

- July 2, 2026: Initial release.
- September 11, 2026: Sends a named User-Agent on API calls (api.useapi.net rejects urllib's default one with HTTP 403), saves encodedImage when fifeUrl is absent, and a failed image download no longer stops the run or leaves a partial file for the next run to skip as "already exists".

"""

import base64
import json
import os
import sys
import time
import urllib.request
import urllib.error

DEFAULT_PROMPTS_FILE = "prompts.json"
DEFAULT_MODEL = "nano-banana-2-lite"
SLEEP_RETRY = 30  # seconds

URL_ACCOUNTS = "https://api.useapi.net/v1/google-flow/accounts"
URL_IMAGES = "https://api.useapi.net/v1/google-flow/images"

# api.useapi.net's Cloudflare edge rejects urllib's default User-Agent (error 1010, HTTP 403).
USER_AGENT = "google-flow-nano-banana-compare.py"


def fetch_accounts(api_token):
    req = urllib.request.Request(
        URL_ACCOUNTS,
        headers={
            "Accept": "application/json",
            "Authorization": f"Bearer {api_token}",
            "User-Agent": USER_AGENT,
        },
    )
    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        print(f"⛔ Error fetching accounts (HTTP {e.code})")
        sys.exit(1)


# Save one generated image. fifeUrl is normally present; when it is absent the image
# arrived inline as base64 in encodedImage. A failure is reported and skipped so the
# remaining images are still saved.
def download_image(image, filename):
    if os.path.exists(filename):
        print(f"⚠️ {filename} already exists. Skipping download.")
        return
    url = image.get("fifeUrl")
    if not url and image.get("encodedImage"):
        print(f"✅ Saving {filename} from encodedImage")
        try:
            data = base64.b64decode(image["encodedImage"])
            with open(filename, "wb") as f:
                f.write(data)
        except Exception as e:
            print(f"⛔ Unable to save {filename}: {e}")
        return
    if not url:
        print(f"🛑 No fifeUrl or encodedImage for {filename}")
        return
    print(f"✅ Downloading {filename}")
    try:
        # Read the whole body before creating the file, so a failed download leaves no
        # partial file behind for the next run to skip as "already exists".
        with urllib.request.urlopen(url) as resp:
            data = resp.read()
        with open(filename, "wb") as f:
            f.write(data)
    except urllib.error.HTTPError as e:
        print(f"⛔ Unable to download {filename} (HTTP {e.code})")
    except Exception as e:
        print(f"⛔ Error during download of {filename}: {e}")


def submit_image(api_token, email, prompt, index):
    model = prompt.get("model", DEFAULT_MODEL)
    body = {"model": model, "email": email, "prompt": prompt["prompt"]}
    for key in ("aspectRatio", "count", "seed"):
        if key in prompt:
            body[key] = prompt[key]

    print(f"🚀 {model} » Prompt #{index} • account {email} …")
    start = time.time()

    while True:
        req = urllib.request.Request(
            URL_IMAGES,
            method="POST",
            data=json.dumps(body).encode(),
            headers={
                "Accept": "application/json",
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_token}",
                "User-Agent": USER_AGENT,
            },
        )
        try:
            with urllib.request.urlopen(req) as resp:
                data = json.loads(resp.read().decode())
                media = data.get("media") or []
                if not media:
                    print(f"🛑 200 OK but no media (moderated?) for {model}")
                for i, m in enumerate(media):
                    gen = (m.get("image") or {}).get("generatedImage") or {}
                    download_image(gen, f"{model}_{i + 1}.jpg")
                print(f"🆗 {model} done ({time.time() - start:.0f} sec)")
                return 200
        except urllib.error.HTTPError as e:
            if e.code in (429, 503):
                print(f"🔄️ Retry on HTTP {e.code}, waiting {SLEEP_RETRY}s")
                time.sleep(SLEEP_RETRY)
                continue
            print(f"🛑 {model} failed with HTTP {e.code}: {e.read().decode()[:200]}")
            return e.code


def main():
    if len(sys.argv) < 3:
        print("Usage: python3 google-flow-nano-banana-compare.py <API_TOKEN> <EMAIL> [PROMPTS_FILE]")
        sys.exit(1)

    api_token, email = sys.argv[1], sys.argv[2]
    prompt_file = sys.argv[3] if len(sys.argv) > 3 else DEFAULT_PROMPTS_FILE

    accounts = fetch_accounts(api_token)
    if email not in accounts:
        print(f"⛔ Account {email} not found. See https://useapi.net/docs/start-here/setup-google-flow")
        sys.exit(1)
    health = accounts[email].get("health")
    if health and health != "OK":
        print(f"⛔ Account {email} health is '{health}'.")
        sys.exit(1)

    with open(prompt_file, encoding="utf-8") as f:
        prompts = json.load(f)
    print(f"Comparing {len(prompts)} model runs …")

    start = time.time()
    for i, prompt in enumerate(prompts, 1):
        submit_image(api_token, email, prompt, i)
    print(f"COMPLETED • elapsed {time.time() - start:.0f} sec")


if __name__ == "__main__":
    main()

File: nano-banana-compare/google-flow-nano-banana-compare.mjs
/*

Script version 1.0, July 2, 2026

Batch-compare Google Flow's Nano Banana image models (2 Lite, 2, Pro) with the useapi.net API 🚀
Reads prompts.json (one entry per model), submits each to the synchronous POST /images endpoint,
and downloads every returned image with a model-labeled filename.
For more details visit https://useapi.net/docs/api-google-flow-v1/post-google-flow-images

Companion tutorial:
  https://useapi.net/docs/articles/google-flow-nano-banana-compare

Installation Instructions:
==========================

You need Node.js v21 or newer installed to run this script. Download and install Node.js from:

- Windows, macOS, Linux: https://nodejs.org/

After installation, verify by running the following command in a terminal:

   node -v

Running the Script:
===================

Usage: node google-flow-nano-banana-compare.mjs <API_TOKEN> <EMAIL> [PROMPTS_FILE]

Replace API_TOKEN with your actual useapi.net API token, see https://useapi.net/docs/start-here/setup-useapi
Replace EMAIL with configured Google Flow email account, see https://useapi.net/docs/start-here/setup-google-flow
If optional PROMPTS_FILE not provided prompts.json will be used.

Example:
--------

node google-flow-nano-banana-compare.mjs user:1234-abcdefhijklmnopqrstuv my@email.com

Changelog:
==========

- July 2, 2026: Initial release.

*/

import fs from 'fs/promises';
import { writeFile } from 'node:fs/promises';
import { Readable } from 'node:stream';

// Constants
const DEFAULT_PROMPTS_FILE = 'prompts.json';
const DEFAULT_MODEL = 'nano-banana-2-lite';
const SLEEP_RETRY = 30 * 1000; // in milliseconds

const urlAccounts = 'https://api.useapi.net/v1/google-flow/accounts';
const urlImages = 'https://api.useapi.net/v1/google-flow/images';

// Utility to sleep for given milliseconds
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
const elapsedTimeSec = (start) => (Date.now() - start) / 1000;

// Function to fetch configured Google Flow API accounts
async function fetchAccounts(apiToken) {
    const response = await fetch(urlAccounts, {
        headers: {
            'Accept': 'application/json',
            'Authorization': `Bearer ${apiToken}`
        }
    });

    if (!response.ok) {
        console.error(`⛔ Error fetching accounts (HTTP ${response.status}): ${response.statusText}`);
        process.exit(1);
    }

    return response.json();
}

// Download a single image from its signed fifeUrl.
async function downloadImage(url, filename) {
    try {
        await fs.access(filename);
        console.log(`⚠️ ${filename} already exists. Skipping download.`);
        return;
    } catch {
        // File does not exist, proceed with downloading
    }

    if (!url) {
        console.error(`🛑 No fifeUrl for ${filename}`);
        return;
    }

    console.log(`✅ Downloading ${filename}`);
    try {
        const imageResponse = await fetch(url);
        if (!imageResponse.ok) {
            console.error(`⛔ Unable to download ${filename} (HTTP ${imageResponse.status})`);
            return;
        }
        await writeFile(filename, Readable.fromWeb(imageResponse.body));
    } catch (err) {
        console.error(`⛔ Error during download: ${err}`);
    }
}

// Submit a single prompt to the synchronous POST /images endpoint and download the results.
async function submitImage(apiToken, email, prompt, index) {
    const model = prompt.model ?? DEFAULT_MODEL;

    // Build the request body.
    const body = { model, email, prompt: prompt.prompt };
    for (const key of ['aspectRatio', 'count', 'seed'])
        if (prompt[key] !== undefined) body[key] = prompt[key];

    console.log(`🚀 ${model} » Prompt #${index} • account ${email} …`);
    const startTime = Date.now();

    while (true) {
        const response = await fetch(urlImages, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${apiToken}`
            },
            body: JSON.stringify(body)
        });

        const responseText = await response.text();

        if (response.status == 200) {
            const media = JSON.parse(responseText)?.media ?? [];

            if (media.length == 0)
                console.error(`🛑 200 OK but no media (moderated?) for ${model}`);

            // count > 1 returns multiple images in the media array.
            for (let i = 0; i < media.length; i++) {
                const img = media[i]?.image?.generatedImage;
                await downloadImage(img?.fifeUrl, `${model}_${i + 1}.jpg`);
            }
            console.log(`🆗 ${model} done (${elapsedTimeSec(startTime)} sec)`);
            return 200;
        }

        switch (response.status) {
            case 429:
            case 503:
                console.log(`🔄️ Retry on HTTP ${response.status}`, responseText);
                await sleep(SLEEP_RETRY);
                break;
            default:
                console.error(`🛑 ${model} failed with HTTP ${response.status}`, responseText);
                return response.status;
        }
    }
}

// Main function
async function main() {
    const apiToken = process.argv[2];
    const email = process.argv[3];
    const promptFile = process.argv[4] || DEFAULT_PROMPTS_FILE;

    if (!apiToken || !email) {
        console.error('Usage: node google-flow-nano-banana-compare.mjs <API_TOKEN> <EMAIL> [PROMPTS_FILE]');
        process.exit(1);
    }

    console.info('Script v1.0 • Node version ' + process.version);

    const accounts = await fetchAccounts(apiToken);

    if (!accounts[email]) {
        console.error(`⛔ Account ${email} not found. Please refer to https://useapi.net/docs/start-here/setup-google-flow`);
        process.exit(1);
    }

    if (accounts[email].health && accounts[email].health !== 'OK') {
        console.error(`⛔ Account ${email} health is '${accounts[email].health}'. Please resolve and update the account, see https://useapi.net/docs/start-here/setup-google-flow`);
        process.exit(1);
    }

    const prompts = JSON.parse(await fs.readFile(promptFile, 'utf8'));
    console.log(`Comparing ${prompts.length} model runs …`);

    const start = new Date();
    for (let i = 0; i < prompts.length; i++)
        await submitImage(apiToken, email, prompts[i], i + 1);

    console.info('COMPLETED', new Date(), '• elapsed', elapsedTimeSec(start), 'sec');
}

main();
